Fix score computation in knowledge_storage to return cosine similarity

chinafzy opened the pull request:
## What’s fixed

1. Collection space set to cosine

When initializing a collection we now explicitly pass `metadata={"hnsw:space": "cosine"}` .
This ensures Chroma computes cosine distance, as recommended in the docs.
``` python
            if self.app:
                self.collection = self.app.get_or_create_collection(
                    name=sanitize_collection_name(collection_name),
                    embedding_function=self.embedder,
                    metadata={"hnsw:space": "cosine"},
                )
```

2. KnowledgeStorage.search now converts the value that Chroma returns (cosine distance) into a true cosine similarity:
`"score": 1 - fetched["distances"][0][i]`
```python
               for i in range(len(fetched["ids"][0])):  # type: ignore
                    result = {
                        "id": fetched["ids"][0][i],  # type: ignore
                        "metadata": fetched["metadatas"][0][i],  # type: ignore
                        "context": fetched["documents"][0][i],  # type: ignore
                        "score": 1 - fetched["distances"][0][i],  # type: ignore
                    }
```

## Why
When a collection is created with metadata={"hnsw:space": "cosine"}, Chroma’s query API returns cosine distance (0 = identical, larger = less similar) rather than cosine similarity.
According to Chroma FAQ “Distances and Similarity” <https://cookbook.chromadb.dev/faq/#distances-and-similarity>, the correct mapping is
```text
similarity = 1 – distance
```

Down-stream code and tests expect a higher score to mean “more relevant”, so the conversion is necessary.

## Impact
+ Search results are now ranked by cosine similarity (0 – 1) instead of the raw distance value.
+ No breaking API changes.

Feel free to let me know if any additional changes are needed!

joaomdmoura commented:
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for PR #3131

## Overview
The modifications in this PR enhance the knowledge storage implementation by improving the calculation of cosine similarity scores and correctly configuring the vector database space. These changes have significant implications for the effectiveness and accuracy of semantic searches, especially when dealing with text embeddings.

## Key Improvements

### 1. Score Computation Change
**Modification**:
```python
- "score": fetched["distances"][0][i],  # type: ignore
+ "score": 1 - fetched["distances"][0][i],  # type: ignore
```
**Analysis**: 
- This alteration effectively transitions from a direct distance measurement to cosine similarity, which is more appropriate for assessing semantic similarity between vectors. 
- Suggestion: It would enhance maintainability to include a comment clarifying this conversion:
```python
# Convert distance to cosine similarity score (1 - distance)
"score": 1 - fetched["distances"][0][i],  # type: ignore
```

### 2. Vector Space Configuration
**Modification**:
```python
metadata={"hnsw:space": "cosine"},
```
**Analysis**: 
- The initialization of the knowledge storage reflects a clear configuration of using cosine similarity, crucial for ensuring accurate representation of semantic relationships.
- Suggestions for improvement:
  1. Implement validation for supported space types to mitigate runtime errors:
  ```python
  SUPPORTED_VECTOR_SPACES = ["cosine", "euclidean", "dot"]

  def initialize_knowledge_storage(self):
      if self.app:
          space_config = {"hnsw:space": "cosine"}
          if space_config["hnsw:space"] not in SUPPORTED_VECTOR_SPACES:
              raise ValueError(f"Unsupported vector space. Must be one of {SUPPORTED_VECTOR_SPACES}")
  ```
  2. Enhance documentation to explain the use of cosine similarity space:
  ```python
  def initialize_knowledge_storage(self):
      """Initialize the knowledge storage with vector database configuration.

      This sets up the collection with cosine similarity space for more accurate
      semantic search results. The preferred use of cosine similarity is due to
      its focus on the angle between vectors rather than their magnitude.
      """
  ```

## Overall Assessment
### Strengths
- The transition to cosine similarity scoring enhances the interpretability and effectiveness of search results.
- Proper initialization with correct vector space configurations ensures better accuracy in retrieval.
- The implementation maintains backward compatibility, safeguarding existing functionality.

### Recommendations for Improvement
- Add documentation for score conversion and method functionalities to ensure code clarity.
- Include validation for vector space types to improve robustness.
- Consider implementing tests specifically targeting the new score computation logic.
- Introduce error handling for unsupported space configurations to ensure system reliability.

### Impact Analysis
- The changes significantly boost the accuracy of semantic searches.
- Aligning with standard practices for similarity metrics enhances user understanding and adaptation.

## Conclusion
The changes implemented in PR #3131 are well-structured and improve the overall functionality and accuracy of the knowledge storage system. The recommendations provided are aimed at further enhancing code maintainability and robustness. Given the sound nature of the current implementation, it is ready for merging following consideration of the suggested improvements.

chinafzy commented:
Thanks for the review! 🙏  
I’ve addressed the feedback and added some clarifications.

### 1. Score computation  
* Added an inline comment explaining that Chroma returns **cosine distance** and we convert it to similarity via `1 - distance`.

### 2. Vector-space setting (`hnsw:space`)  
All CrewAI text embeddings are **L2-normalized**, so **cosine distance** (or its equivalent `ip`) is the only metric that preserves semantic ordering for our use-case.  
According to the official Chroma docs — *Collections ▸ Configure ▸ HNSW Index Configuration*:

| key | allowed values |
|-----|----------------|
| `hnsw:space` | `l2`, `cosine`, `ip` |

<https://docs.trychroma.com/docs/collections/configure#hnsw-index-configuration>

`euclidean` / `dot` are **not** valid values.  


I therefore keep the collection fixed to `"cosine"` and documented this assumption in code:

```python
# “cosine” is the only supported metric in CrewAI for now; do NOT change.
_VECTOR_SPACE = "cosine"
...
metadata={"hnsw:space": _VECTOR_SPACE},
```

If we ever ingest non-normalized vectors (e.g. image embeddings) we’ll introduce a configurable parameter and matching tests in a future PR.

### 3. No functional changes beyond docs  
Only comments/constant extraction were added; logic remains identical and all tests pass locally.

Let me know if anything else is needed — otherwise please **Approve & run the workflows** when convenient. Thanks!

greysonlalonde commented:
Fixed in #3374
